src/geometry/mixed_lines_set.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright (c) 2024 UltiMaker
+// CuraEngine is released under the terms of the AGPLv3 or higher.
+
+#include "geometry/mixed_lines_set.h"
+
+#include <numeric>
+
+#include "geometry/open_polyline.h"
+#include "geometry/polygon.h"
+#include "geometry/shape.h"
+
+
+namespace cura
+{
+
+Shape MixedLinesSet::offset(coord_t distance, ClipperLib::JoinType join_type, double miter_limit) const
+{
+    if (distance == 0)
+    {
+        // Return a shape that contains only actual polygons
+        Shape result;
+
+        for (const PolylinePtr& line : (*this))
+        {
+            if (const std::shared_ptr<const Polygon> polygon = dynamic_pointer_cast<const Polygon>(line))
```

Comment:
Like different types of cheese you mean, an 'acquired taste' 😆 

I'm a bit on the fence... I feel like part of the point of polymorphism is that you shouldn't need to do things like this. -- I need to think about it a bit more.  But we can also move ahead since, like you said, a good point of the new chosen architecture is that it has a well defined interface anyway.

The Visitor pattern is just a suggestion. It might be overkill anyway. (And I'm not sold on any naming either -- if it confuses people it's not maintainable after all!)